src/converse-httpFileUpload.js

```diff
@@ -0,0 +1,269 @@
+                        ready = true;
+                        return;
+                    }
+                    this.discoverUploadService();
+                },
+                
+                /**
+                * Discover upload service for http upload.
+                *
+                */
+                discoverUploadService () {
+                    var self = this;
+                    console.log('discover http upload service');
+                    connection.disco.items(domain, null, function(items) {
+                        var childs = items.getElementsByTagName('item');
+                        for(var i = 0; i < childs.length; i++){
+                            var jid = childs[i].attributes.jid.value; 
+                            if (ready) {
+                                // abort, because we already found a service
+                                return false;
+                            }
+                            self.queryItemForUploadService(jid);
+                        }
+                    });
+                },
+                
+                /**
+                 * Query item for upload service.
+                 *
+                 * @param {String} jid of the logged-in user
+                 * @param {Function} cb Callback on success
+                */
+                queryItemForUploadService (jid) {
+                    var self = this;
+                    console.log('query ' + jid + ' for upload service');
+                
+                    connection.disco.info(jid, null, function(info) {
+                        var httpUploadFeature;
+                        var temp = info.getElementsByTagName('feature');
+                        for(var i = 0; i < temp.length; i++){
+                            var feature = temp[i].attributes.var;
+                            if(feature != undefined && feature.value === Strophe.NS.HTTPUPLOAD){
+                                requestSlotUrl = jid;
+                                ready = true;
+                                self.sendFile();
+                            }
+                        }
+                    });
+                },
+
+                /**
+                 * Saves the file the user has picked.
+                 * 
+                 * @param {*} file the name of the file the user has picked.
+                 * @param {*} chatBox the chatbox from which the user initiated the file-upload
+                 */
+                setFile (file1, chatBox1){
+                    file = file1;
+                    chatBox = chatBox1;
+                    this.sendFile();
+                },
+                
+                /**
+                * Upload file.
+                * Waits till the Upload-Service is discovered and till the user has picked a file.
+                *
+                */
+                sendFile () {
+                    var self = this;
```

Comment:
Inasmuch this method is needed, it should go onto the `_converse.ChatBox` model.

That said, given what I've seen above, I suspect that a rewrite of the rest of the methods in this file might be necessary.